blockindex/liquidstaking_cache.go

```diff
@@ -0,0 +1,165 @@
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/iotexproject/iotex-core/db/batch"
+)
+
+type (
+	liquidStakingCache struct {
+		idBucketMap           map[uint64]*BucketInfo     // map[token]BucketInfo
+		candidateBucketMap    map[string]map[uint64]bool // map[candidate]bucket
+		idBucketTypeMap       map[uint64]*BucketType     // map[token]BucketType
+		propertyBucketTypeMap map[int64]map[int64]uint64 // map[amount][duration]index
+		height                uint64
+	}
+)
+
+func newLiquidStakingCache() *liquidStakingCache {
+	return &liquidStakingCache{
+		idBucketMap:           make(map[uint64]*BucketInfo),
+		idBucketTypeMap:       make(map[uint64]*BucketType),
+		propertyBucketTypeMap: make(map[int64]map[int64]uint64),
+		candidateBucketMap:    make(map[string]map[uint64]bool),
+	}
+}
+
+func (s *liquidStakingCache) writeBatch(b batch.KVStoreBatch) error {
+	for i := 0; i < b.Size(); i++ {
+		write, err := b.Entry(i)
+		if err != nil {
+			return err
+		}
+		switch write.Namespace() {
+		case _liquidStakingBucketInfoNS:
+			if write.WriteType() == batch.Put {
+				var bi BucketInfo
+				if err = bi.deserialize(write.Value()); err != nil {
+					return err
+				}
+				id := deserializeUint64(write.Key())
+				s.putBucketInfo(id, &bi)
+			} else if write.WriteType() == batch.Delete {
+				id := deserializeUint64(write.Key())
+				s.deleteBucketInfo(id)
+			}
+		case _liquidStakingBucketTypeNS:
+			if write.WriteType() == batch.Put {
+				var bt BucketType
+				if err = bt.deserialize(write.Value()); err != nil {
+					return err
+				}
+				id := deserializeUint64(write.Key())
+				s.putBucketType(id, &bt)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *liquidStakingCache) putHeight(h uint64) {
+	s.height = h
+}
+
+func (s *liquidStakingCache) getHeight() uint64 {
+	return s.height
+}
+
+func (s *liquidStakingCache) putBucketType(id uint64, bt *BucketType) {
+	amount := bt.Amount.Int64()
+	s.idBucketTypeMap[id] = bt
+	m, ok := s.propertyBucketTypeMap[amount]
+	if !ok {
+		s.propertyBucketTypeMap[amount] = make(map[int64]uint64)
+		m = s.propertyBucketTypeMap[amount]
+	}
+	m[int64(bt.Duration)] = id
+}
+
+func (s *liquidStakingCache) putBucketInfo(id uint64, bi *BucketInfo) {
+	s.idBucketMap[id] = bi
+	if _, ok := s.candidateBucketMap[bi.Delegate]; !ok {
+		s.candidateBucketMap[bi.Delegate] = make(map[uint64]bool)
+	}
+	s.candidateBucketMap[bi.Delegate][id] = true
+}
+
+func (s *liquidStakingCache) deleteBucketInfo(id uint64) {
+	bi, ok := s.idBucketMap[id]
+	if !ok {
+		return
+	}
+	delete(s.idBucketMap, id)
+	if _, ok := s.candidateBucketMap[bi.Delegate]; !ok {
+		return
+	}
+	delete(s.candidateBucketMap[bi.Delegate], id)
+}
+
+func (s *liquidStakingCache) markDeleteBucketInfo(id uint64) {
```

Comment:
what's the diff with above func?

Reply:
it sets value to nil instead of deleting key from map, to detect if it's deleted. but i have refactored these code, adding a delta struct, y can just forget it